p2p/src/network/kad/bootstrap/p2p_network_kad_bootstrap_reducer.rs

```diff
@@ -45,21 +45,23 @@ fn prepare_next_request(
 
 impl P2pNetworkKadBootstrapState {
     pub fn reducer<Action, State>(
-        mut state_context: Substate<Action, State, Self>,
+        mut state_context: Substate<Action, State, P2pState>,
         action: ActionWithMeta<&P2pNetworkKadBootstrapAction>,
         filter_addrs: bool,
     ) -> Result<(), String>
     where
-        State: SubstateAccess<Self> + SubstateAccess<P2pState> + SubstateAccess<P2pNetworkKadState>,
+        State: crate::P2pStateTrait,
         Action: crate::P2pActionTrait<State>,
     {
         let (action, meta) = action.split();
-        let discovery_state: &P2pNetworkKadState = state_context.get_state().substate()?;
-        let routing_table = &discovery_state.routing_table;
-        let bootstrap_state = state_context.get_substate()?;
 
         match action {
             P2pNetworkKadBootstrapAction::CreateRequests => {
+                let discovery_state: &P2pNetworkKadState =
+                    state_context.get_substate()?.substate()?;
```

Comment:
Oh I didn't notice the routing table was part of the substate, this is better then. I wonder why it was being passed from the caller before.

Reply:
nevermind, I see this is reducing over the parent state (that works too).

Reply:
One concern I have is that the way it is defiend now, allows for this reducer to mutate the outer state, which is not ideal (the code doesn't do that, but it is not obvious from just the signature).

Reply:
(btw, not a blocker, we can figure it out later)

Reply:
This is also case for reducer on requests and streams.